rust/rope/src/spans.rs

```diff
@@ -300,6 +300,40 @@ impl<T: Clone> Spans<T> {
         }
         *self = b.build();
     }
+
+    /// Deletes all spans that intersect with `interval`.
+    pub fn delete_intersecting(&mut self, interval: Interval) {
+        let mut cursor = Cursor::new(self, interval.start());
+
+        while cursor.pos() > interval.start() && cursor.prev_leaf().is_some() {}
```

Comment:
this is a strange pattern, and I'm not totally sure what you're trying to do. 

This _is_ a tricky problem for interval trees. For instance, there may be an interval in the first leaf that spans the entire tree needs to be removed correctly. As a simple example, the following test case fails here:

```rust
    #[test]
    fn delete_intersecting_big_at_start() {
        let mut sb = SpansBuilder::new(10);
        sb.add_span(0..10, 0);

        let mut spans = sb.build();
        assert_eq!(spans.iter().count(), 1);

        spans.delete_intersecting(1..3);
        assert_eq!(spans.iter().count(), 0);
    }
```

In any case, happy to chat about this!

Reply:
Looking at this again I am also not sure why I iterated over the leaves. 

I fixed the logic in that function, but it is a bit hairy now. Maybe any suggestions on a cleaner way to write this?